src/lib/eliom_notif.server.ml

```diff
@@ -0,0 +1,197 @@
+      then Weak_tbl.add wt v;
+      Lwt.return ()
+    )
+
+    let iter =
+      let iter (f : Weak_tbl.data -> unit Lwt.t) wt : unit =
+	Weak_tbl.iter
+	  (fun data -> Lwt.async (fun () -> f data))
+	  wt
+      in
+      fun f key -> async_locked (fun () ->
+	let () =
+	  try
+	    let wt = Notif_hashtbl.find tbl key in
+	    let g data = match data with
+              | None ->
+		Weak_tbl.remove wt data;
+		remove_if_empty wt key;
+		Lwt.return ()
+              | Some v ->
+		f v;
+		Lwt.return ()
+	    in
+            iter g wt;
+	  with Not_found -> ()
+	in
+	Lwt.return ()
+      )
+  end
+
+  let identity_r : (A.identity * notification_react) option Eliom_reference.eref =
+    Eliom_reference.eref
+      ~scope:Eliom_common.default_process_scope
+      None
+
+  (* notif_e consists in a server side react event,
+     its client side counterpart,
+     and the server side function to trigger it. *)
+  let notif_e : notification_react option Eliom_reference.eref =
+    Eliom_reference.eref
+      ~scope:Eliom_common.default_process_scope
+      None
+
+  let of_option = function
+    | Some x -> x
+    | None -> assert false
+
+  let set_identity identity =
+    (* For each tab connected to the app,
+       we keep a pointer to (identity, notif_ev) option in process state,
+       because the table resourceid -> (identity, notif_ev) option
+       is weak.
+    *)
+    Eliom_reference.get notif_e >>= fun notif_o ->
+    Eliom_reference.set identity_r (Some (identity, of_option notif_o))
+
+  let set_notif_e () =
+    let notif =
+      let e, send_e = React.E.create () in
+      let client_ev = Eliom_react.Down.of_react
+      (*VVV If we add throttling, some events may be lost
+            even if buffer size is not 1 :O *)
+	~size: 100 (*VVV ? *)
+	~scope:Eliom_common.default_process_scope
+	e
+      in
+      (client_ev, send_e)
+    in
+    Eliom_reference.set notif_e (Some notif)
+
+  let set_current_identity () =
+    A.get_identity () >>= fun identity ->
+    set_identity identity
+
+  let listen (key : A.key) = Lwt.async (fun () ->
+    set_notif_e () >>= fun () ->
+    set_current_identity () >>= fun () ->
+    Eliom_reference.get identity_r >>= fun identity ->
+    I.add identity key;
+    Lwt.return ()
+  )
+
+  let unlisten (id : A.key) = Lwt.async (fun () ->
+    Eliom_reference.get identity_r >>= fun identity ->
+    I.remove identity id;
+    Lwt.return ()
+  )
+
+  let notify ?(notforme = false) key content_gen =
+    let f = fun (identity, ((_, send_e) as notif)) ->
+      Eliom_reference.get notif_e >>= fun notif_o ->
+      if notforme && notif == (of_option notif_o) then
+	Lwt.return ()
+      else
+        content_gen identity >>= fun content -> match content with
+        | Some content -> send_e (key, content); Lwt.return ()
+        | None -> Lwt.return ()
+    in
+    (* on all tabs registered on this data *)
+    I.iter f key
+
+  let client_ev () =
+    let (ev, _) = Lwt_main.run (
```

Comment:
I am uncomfortable with this `Lwt_main.run`. Can `client_ev` return an Lwt thread instead?